### PR TITLE
Zqs 870/track backend usage

### PR DESCRIPTION
**Pull request by AthenaCaesura**
Created MeasurementTrackingBackend class which allows one to keep track of all the measurement data collected in an experiment in an external json file. Namely, we keep track of num_shots, bitstrings, counts for each measurement outcome, the implemented circuit, and the number of gates.

This generalized code written in qe-qiskit which only tracked backends for qiskit. This code expands the scope to any type of backend.

The measurement tracking backend class is triggered when the user puts the flag "track_backend_usage" : True in the workflow. The class wraps around an inner_backend and writes all the results of measurements to "raw_measurement_data.json" after each experiment. 

Testing for measurement tracking backend to be added in an upcoming story.

### CODE REVIEW

**Review by mstechly — CHANGES_REQUESTED** [review]

File: .coveragerc
Comment (AthenaCaesura) [reply .coveragerc]: I found another copy of .coveragerc in the main branch and put an identical file in my branch. Hope that's ok.

File: src/python/zquantum/core/trackers.py
Comment (AthenaCaesura) [reply src/python/zquantum/core/trackers.py]: Done. :)

File: src/python/zquantum/core/trackers.py
Comment (AthenaCaesura) [reply src/python/zquantum/core/trackers.py]: I put the comment in the docstring of all of the methods in steps where MeasurementTrackingBackend could be called. If methods have no docstrings, I just put the comment in the module docstring (i.e. `measurements.py`).

File: src/python/zquantum/core/trackers.py
Comment (alexjuda) [inline src/python/zquantum/core/trackers.py:139]: Do I understand correctly that the motivation for this method was to let the user to store the data under a different file name? WDYT about removing this method and allowing the user to pass a file object to `save_raw_measurement_data`?

File: steps/evaluation.py
Comment (alexjuda) [inline steps/evaluation.py:83]: Shouldn't this follow invocation of `backend.save_raw_measurement_data()`? If so, can you help me find it?

Note that this comment will be out of date if you decide to merge `save...` and `rename...` methods.

File: tests/zquantum/core/trackers_test.py
Comment (alexjuda) [inline tests/zquantum/core/trackers_test.py:52]: If we can pass `n_samples` to `run_circuit_and_measure`, what's the reason to mutate the backend on line above it?

File: tests/zquantum/core/trackers_test.py
Comment (alexjuda) [inline tests/zquantum/core/trackers_test.py:53]: 1. If you don't wrap `open(...)` in a context manager you're responsible for closing the file pointer on your own. I can't see the `f.close()` in this test, so I'd expect FS errors due to removal of an dangling/opened file at the test cleanup. Is there a reason not to use `with open(...) as f`?
2. If this test fails for some reason we end up with temporary files cluttering the current PWD. WDYT about using a [tempfile]()https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) or a [tempdir](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory)? You can see how we use that in [SDK's tests](https://github.com/zapatacomputing/orquestra-sdk/blob/a9b2f1d94d0f7671c7bbb357acfde1ffb8dbb9ba/tests/v2/test_dispatch.py#L10-L46)

**Review by alexjuda — COMMENTED** [review]
I've added a few questions/suggestions :) 

Side question: we still have ["acceptance tests"](https://github.com/zapatacomputing/z-quantum-core/tree/main/tests/acceptance_tests) that only contain tests for out steps. Wanna add tests for backend tracker there, too?

**Comment by mstechly** [comment]
@alexjuda regarding acceptance tests – I don't think it's necessary. Since we'll be deleting them in near future, I don't think it's worth spending time on them.

**Review by mstechly — APPROVED** [review]
Looks good to me!
Just please add `.coveragerc` again ;)
Waiting for approval from @alexjuda

**Pull request merged**
